// @coolcinema/cli@4.0.34/dist/handlers/ingress.handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IngressHandler = void 0;
const zod_1 = require("zod");
const base_handler_1 = require("./base.handler");
const IngressRuleSchema = zod_1.z
    .object({
    port: zod_1.z.number(),
    host: zod_1.z.string().optional(),
    path: zod_1.z.string().default("/"),
})
    .strict();
const IngressSectionSchema = zod_1.z.record(zod_1.z.string(), IngressRuleSchema);
class IngressHandler extends base_handler_1.BaseHandler {
    constructor() {
        super(...arguments);
        this.schema = IngressSectionSchema;
        this.defaults = {
            main: {
                port: 3000,
                path: "/",
            },
        };
    }
    async processItem(ctx, key, item, result) {
        const prefix = key === "main" ? "" : `${key}.`;
        const host = item.host || `${prefix}${ctx.serviceSlug}.coolcinema.local`;
        result.appConfig.ingress.push({
            name: key,
            port: item.port,
            host: host,
            path: item.path,
        });
        result.appConfig.ports.push({
            name: `http-${key}`,
            port: item.port,
            protocol: "TCP",
        });
        result.registryData[key] = item;
    }
}
exports.IngressHandler = IngressHandler;
