// @coolcinema/cli@4.0.27/dist/services/registry.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegistryService = void 0;
const config_1 = require("../config");
class RegistryService {
    constructor(gh) {
        this.gh = gh;
        this.data = { services: {} };
    }
    async fetch() {
        const content = await this.gh.getFileContent(config_1.CONFIG.PATHS.REGISTRY_JSON);
        if (content) {
            this.data = JSON.parse(content);
        }
    }
    getServices() {
        return this.data.services || {};
    }
    updateService(slug, metadata, interfaces) {
        this.data.services[slug] = {
            host: slug,
            ...metadata,
            ...interfaces,
        };
    }
    loadFromLocal() {
        try {
            // Пытаемся найти пакет contracts
            // Путь может отличаться в зависимости от структуры проекта, но require найдет его
            // @ts-ignore
            const localRegistry = require("@coolcinema/contracts/dist/registry.json");
            this.data = { services: localRegistry.services };
        }
        catch (e) {
            console.warn("⚠️  Could not load local registry. Make sure @coolcinema/contracts is installed.");
        }
    }
    getArtifact() {
        return {
            path: config_1.CONFIG.PATHS.REGISTRY_JSON,
            content: JSON.stringify(this.data, null, 2),
        };
    }
}
exports.RegistryService = RegistryService;
