// @coolcinema/cli@4.0.28/dist/utils/error-handler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorHandler = void 0;
const chalk_1 = __importDefault(require("chalk"));
const zod_1 = require("zod");
const IndentationHinter = (issue, handlers) => {
    if (issue.code !== zod_1.ZodIssueCode.unrecognized_keys)
        return null;
    if (issue.path.length !== 1 || issue.path[0] !== "metadata")
        return null;
    for (const key of issue.keys) {
        if (handlers[key]) {
            return `Key '${key}' found inside 'metadata'. It should be at the ROOT level. Check indentation!`;
        }
    }
    return null;
};
const activeHinters = [IndentationHinter];
class ErrorHandler {
    static handle(err, handlers = {}) {
        if (err instanceof zod_1.z.ZodError) {
            this.handleZodError(err, handlers);
        }
        else {
            this.handleGenericError(err);
        }
        process.exit(1);
    }
    static handleZodError(err, handlers) {
        console.error(chalk_1.default.red("❌ Manifest Validation Failed:"));
        for (const issue of err.issues) {
            this.printIssue(issue);
            this.printHints(issue, handlers);
        }
    }
    static handleGenericError(err) {
        console.error(chalk_1.default.red("❌ Error:"), err.message || err);
    }
    static printIssue(issue) {
        const pathStr = issue.path.join(".");
        console.error(chalk_1.default.yellow(`- [${pathStr}]: ${issue.message}`));
    }
    static printHints(issue, handlers) {
        for (const hinter of activeHinters) {
            const hint = hinter(issue, handlers);
            if (hint) {
                console.log(chalk_1.default.cyan(`  💡 Tip: ${hint}`));
            }
        }
    }
}
exports.ErrorHandler = ErrorHandler;
