// @coolcinema/foundation@2.0.47/dist/openapi/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRoute = createRoute;
function createRoute(options) {
    const status = options.status || 200;
    const operation = {
        summary: options.summary,
        description: options.description,
        tags: options.tags,
        responses: {
            [status]: {
                description: "Success",
                content: {
                    "application/json": { schema: options.response },
                },
            },
        },
    };
    if (options.params || options.query) {
        operation.requestParams = {};
        if (options.params)
            operation.requestParams.path = options.params;
        if (options.query)
            operation.requestParams.query = options.query;
    }
    if (options.body) {
        operation.requestBody = {
            content: {
                "application/json": { schema: options.body },
            },
        };
    }
    return {
        path: options.path,
        method: options.method,
        config: operation,
        Handler: {},
    };
}
