// @coolcinema/cli@4.0.15/dist/handlers/events.handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsHandler = void 0;
const zod_1 = require("zod");
const base_handler_1 = require("./base.handler");
const config_1 = require("../config");
const EventsInterfaceSchema = zod_1.z
    .object({
    spec: zod_1.z.string(),
})
    .strict();
class EventsHandler extends base_handler_1.BaseHandler {
    constructor() {
        super(...arguments);
        this.schema = EventsInterfaceSchema;
        this.defaults = {
            spec: "src/contracts/events/asyncapi.yaml",
        };
    }
    async push(ctx, rawConfig) {
        const config = rawConfig;
        const result = {
            registryData: {},
            appConfig: { ports: [], ingress: [], env: {} },
        };
        console.log(`[DEBUG] EventsHandler processing spec:`, config.spec);
        const content = await ctx.readFile(config.spec);
        const remoteFileName = `${ctx.serviceSlug}.yaml`;
        const remotePath = `${config_1.CONFIG.PATHS.CONTRACTS_ROOT}/events/${remoteFileName}`;
        ctx.uploadFile(remotePath, content);
        result.registryData = {
            spec: `events/${remoteFileName}`,
            hasEvents: true,
        };
        return result;
    }
    async processItem() { }
}
exports.EventsHandler = EventsHandler;
