// @coolcinema/cli@4.0.34/dist/services/config-manager.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigManager = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const yaml = __importStar(require("js-yaml"));
const config_1 = require("../config");
const chalk_1 = __importDefault(require("chalk"));
class ConfigManager {
    constructor(rootDir) {
        this.rootDir = rootDir;
        this.backupPath = null;
        this.configPath = path.resolve(rootDir, config_1.CONFIG.PATHS.BUF.WORK);
    }
    prepare() {
        if (!fs.existsSync(this.configPath)) {
            console.error(chalk_1.default.red(`❌ ${config_1.CONFIG.PATHS.BUF.WORK} not found.`));
            return false;
        }
        try {
            const originalContent = fs.readFileSync(this.configPath, "utf8");
            const config = yaml.load(originalContent);
            if (config && Array.isArray(config.directories)) {
                const validDirectories = config.directories.filter((dir) => this.hasProtoFiles(path.resolve(this.rootDir, dir)));
                if (validDirectories.length !== config.directories.length) {
                    console.log(chalk_1.default.gray(`🔧 Excluding empty directories from build...`));
                    this.backupPath = `${this.configPath}.bak`;
                    fs.writeFileSync(this.backupPath, originalContent);
                    config.directories = validDirectories;
                    fs.writeFileSync(this.configPath, yaml.dump(config));
                }
            }
            return true;
        }
        catch (e) {
            console.error(chalk_1.default.red("❌ Failed to parse workspace config."));
            return false;
        }
    }
    restore() {
        if (this.backupPath && fs.existsSync(this.backupPath)) {
            fs.writeFileSync(this.configPath, fs.readFileSync(this.backupPath));
            fs.unlinkSync(this.backupPath);
        }
    }
    hasProtoFiles(dir) {
        if (!fs.existsSync(dir))
            return false;
        try {
            const files = fs.readdirSync(dir, { withFileTypes: true });
            for (const file of files) {
                if (file.isDirectory()) {
                    if (this.hasProtoFiles(path.join(dir, file.name)))
                        return true;
                }
                else if (file.name.endsWith(".proto")) {
                    return true;
                }
            }
        }
        catch (e) {
            return false;
        }
        return false;
    }
}
exports.ConfigManager = ConfigManager;
