// @coolcinema/cli@4.0.34/dist/services/manifest.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManifestService = void 0;
const fs = __importStar(require("fs"));
const yaml = __importStar(require("js-yaml"));
const base_1 = require("../schemas/base");
class ManifestService {
    constructor(handlers) {
        this.handlers = handlers;
    }
    load(filePath) {
        if (!fs.existsSync(filePath)) {
            throw new Error("Manifest not found");
        }
        let raw;
        try {
            raw = yaml.load(fs.readFileSync(filePath, "utf8"));
        }
        catch (e) {
            throw new Error(`Invalid YAML: ${e.message}`);
        }
        const Schema = this.createFullSchema().strict();
        return Schema.parse(raw);
    }
    createTemplate(serviceName, slug) {
        const base = base_1.BaseManifestSchema.parse({
            metadata: {
                name: serviceName,
                slug: slug,
            },
        });
        const templateObj = { ...base };
        for (const [key, module] of Object.entries(this.handlers)) {
            if (module.defaults) {
                templateObj[key] = module.defaults;
            }
        }
        return yaml.dump(templateObj, {
            indent: 2,
            lineWidth: -1,
            noRefs: true,
        });
    }
    save(filePath, content, options = {}) {
        if (fs.existsSync(filePath) && !options.overwrite) {
            throw new Error(`File already exists: ${filePath}`);
        }
        fs.writeFileSync(filePath, content);
    }
    createFullSchema() {
        const extensions = {};
        for (const [key, module] of Object.entries(this.handlers)) {
            extensions[key] = module.schema.optional();
        }
        return base_1.BaseManifestSchema.extend(extensions);
    }
}
exports.ManifestService = ManifestService;
