// @coolcinema/foundation@2.0.46/dist/grpc/nest/module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var GrpcModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcModule = void 0;
const common_1 = require("@nestjs/common");
const factory_1 = require("../nice/factory");
let GrpcModule = GrpcModule_1 = class GrpcModule {
    static forRoot(registry, definitions) {
        const clients = (0, factory_1.createClients)(registry, definitions);
        const providers = Object.entries(clients).map(([name, client]) => ({
            provide: `${name}Service`,
            useValue: client,
        }));
        return {
            module: GrpcModule_1,
            providers: providers,
            exports: providers.map((p) => p.provide),
        };
    }
};
exports.GrpcModule = GrpcModule;
exports.GrpcModule = GrpcModule = GrpcModule_1 = __decorate([
    (0, common_1.Global)(),
    (0, common_1.Module)({})
], GrpcModule);
