// @coolcinema/cli@4.0.22/dist/services/github.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitHubService = void 0;
const rest_1 = require("@octokit/rest");
const config_1 = require("../config");
class GitHubService {
    constructor() {
        if (!config_1.CONFIG.GITHUB.TOKEN)
            throw new Error("Missing COOLCINEMA_GH_PKG_TOKEN");
        this.octokit = new rest_1.Octokit({ auth: config_1.CONFIG.GITHUB.TOKEN });
    }
    async getFileContent(path) {
        try {
            const { data } = await this.octokit.rest.repos.getContent({
                owner: config_1.CONFIG.GITHUB.OWNER,
                repo: config_1.CONFIG.GITHUB.REPO,
                path,
                ref: config_1.CONFIG.GITHUB.BRANCH,
            });
            if (!Array.isArray(data) && "content" in data) {
                return Buffer.from(data.content, "base64").toString("utf8");
            }
        }
        catch (e) {
            /* ignore 404 */
        }
        return null;
    }
    async createAtomicCommit(files, message) {
        const { OWNER, REPO, BRANCH } = config_1.CONFIG.GITHUB;
        const { data: ref } = await this.octokit.rest.git.getRef({
            owner: OWNER,
            repo: REPO,
            ref: `heads/${BRANCH}`,
        });
        const latestCommitSha = ref.object.sha;
        const { data: commit } = await this.octokit.rest.git.getCommit({
            owner: OWNER,
            repo: REPO,
            commit_sha: latestCommitSha,
        });
        const baseTreeSha = commit.tree.sha;
        const treeItems = [];
        for (const file of files) {
            const { data: blob } = await this.octokit.rest.git.createBlob({
                owner: OWNER,
                repo: REPO,
                content: file.content,
                encoding: "utf-8",
            });
            treeItems.push({
                path: file.path,
                mode: "100644",
                type: "blob",
                sha: blob.sha,
            });
        }
        const { data: newTree } = await this.octokit.rest.git.createTree({
            owner: OWNER,
            repo: REPO,
            base_tree: baseTreeSha,
            tree: treeItems,
        });
        const { data: newCommit } = await this.octokit.rest.git.createCommit({
            owner: OWNER,
            repo: REPO,
            message,
            tree: newTree.sha,
            parents: [latestCommitSha],
        });
        await this.octokit.rest.git.updateRef({
            owner: OWNER,
            repo: REPO,
            ref: `heads/${BRANCH}`,
            sha: newCommit.sha,
        });
        return newCommit.sha;
    }
}
exports.GitHubService = GitHubService;
