// @coolcinema/cli@4.0.22/dist/handlers/grpc.handler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcHandler = void 0;
const path = __importStar(require("path"));
const zod_1 = require("zod");
const base_handler_1 = require("./base.handler");
const config_1 = require("../config");
const GrpcInterfaceSchema = zod_1.z
    .object({
    files: zod_1.z.array(zod_1.z.string()),
    port: zod_1.z.number(),
})
    .strict();
const GrpcSectionSchema = zod_1.z.record(zod_1.z.string(), GrpcInterfaceSchema);
class GrpcHandler extends base_handler_1.BaseHandler {
    constructor() {
        super(...arguments);
        this.schema = GrpcSectionSchema;
        this.defaults = {
            main: {
                files: ["src/contracts/grpc/service.proto"],
                port: 5000,
            },
        };
    }
    async processItem(ctx, key, item, result) {
        const uploadedFiles = [];
        for (const filePath of item.files) {
            const content = await ctx.readFile(filePath);
            const fileName = `${ctx.serviceSlug}_${path.basename(filePath)}`;
            const remotePath = `${config_1.CONFIG.PATHS.CONTRACTS_ROOT}/protos/${fileName}`;
            ctx.uploadFile(remotePath, content);
            uploadedFiles.push(`protos/${fileName}`);
        }
        result.registryData[key] = {
            files: uploadedFiles,
            port: item.port,
        };
        result.appConfig.ports.push({
            name: "grpc",
            port: item.port,
            protocol: "TCP",
        });
    }
}
exports.GrpcHandler = GrpcHandler;
