// @coolcinema/foundation@2.0.47/dist/events/nest/module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var EventsModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsModule = void 0;
const common_1 = require("@nestjs/common");
const amqp_1 = require("../amqp");
const factory_1 = require("../factory");
let EventsModule = EventsModule_1 = class EventsModule {
    static register(options) {
        const transportProvider = {
            provide: "AMQP_TRANSPORT",
            useFactory: () => new amqp_1.AmqpTransport(options.registry),
        };
        const publisherProvider = {
            provide: options.name,
            useFactory: (transport) => {
                return (0, factory_1.createPublisher)(transport, options.definitions, options.prefix);
            },
            inject: ["AMQP_TRANSPORT"],
        };
        return {
            module: EventsModule_1,
            providers: [transportProvider, publisherProvider],
            exports: [publisherProvider],
        };
    }
};
exports.EventsModule = EventsModule;
exports.EventsModule = EventsModule = EventsModule_1 = __decorate([
    (0, common_1.Global)(),
    (0, common_1.Module)({})
], EventsModule);
