// @coolcinema/cli@4.0.28/dist/services/infra.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfraService = void 0;
const config_1 = require("../config");
class InfraService {
    constructor() {
        this.config = {
            ports: [],
            ingress: [],
            env: {},
        };
    }
    add(partial) {
        if (partial.ports)
            this.config.ports?.push(...partial.ports);
        if (partial.ingress)
            this.config.ingress?.push(...partial.ingress);
        if (partial.env)
            Object.assign(this.config.env, partial.env);
    }
    createArtifact(metadata) {
        const uniquePorts = this.config.ports?.filter((port, index, self) => index === self.findIndex((p) => p.port === port.port)) || [];
        const values = {
            fullnameOverride: metadata.slug,
            image: {
                repository: `ghcr.io/coolcinema/${metadata.slug}`,
            },
            service: {
                ports: uniquePorts,
            },
            ingress: this.config.ingress || [],
            env: this.config.env,
        };
        return {
            path: `${config_1.CONFIG.PATHS.APPS_DIR}/${metadata.slug}${config_1.CONFIG.PATHS.APP_EXT}`,
            content: JSON.stringify(values, null, 2),
        };
    }
}
exports.InfraService = InfraService;
