// @coolcinema/cli@4.0.26/dist/commands/gen-http.command.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenHttpCommand = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const child_process_1 = require("child_process");
const config_1 = require("../config");
const chalk_1 = __importDefault(require("chalk"));
class GenHttpCommand {
    getContractPath() {
        const srcDir = config_1.CONFIG.PATHS.LOCAL_CONTRACTS.HTTP;
        const absPath = path.resolve(srcDir);
        if (!fs.existsSync(absPath))
            return null;
        const files = fs.readdirSync(absPath).filter((f) => f.endsWith(".proto"));
        if (files.length === 0)
            return null;
        return srcDir;
    }
    async postProcess() {
        const rootDir = process.cwd();
        const specOutDir = path.join(rootDir, config_1.CONFIG.PATHS.LOCAL_GEN.HTTP_SPEC);
        const clientOutDir = path.join(rootDir, config_1.CONFIG.PATHS.LOCAL_GEN.HTTP);
        if (!fs.existsSync(specOutDir))
            return;
        // Ищем файлы рекурсивно (Native Recursive Search)
        const specFiles = this.findSpecs(specOutDir);
        if (specFiles.length === 0)
            return;
        if (!fs.existsSync(clientOutDir))
            fs.mkdirSync(clientOutDir, { recursive: true });
        for (const input of specFiles) {
            const relativePath = path.relative(specOutDir, input);
            const output = path.join(clientOutDir, relativePath.replace(/\.(json|yaml)$/, ".ts"));
            const outputDir = path.dirname(output);
            if (!fs.existsSync(outputDir)) {
                fs.mkdirSync(outputDir, { recursive: true });
            }
            try {
                (0, child_process_1.execSync)(`pnpm exec openapi-typescript "${input}" -o "${output}"`, {
                    stdio: "inherit",
                });
            }
            catch (e) {
                console.error(`❌ Client generation failed for ${relativePath}`);
            }
        }
        console.log("✅ HTTP client generated.");
    }
    async execute() {
        const path = this.getContractPath();
        if (!path) {
            console.log(chalk_1.default.gray(`Skipping HTTP: no .proto files`));
            return;
        }
        try {
            console.log("🔨 Generating HTTP spec...");
            (0, child_process_1.execSync)(`pnpm exec buf generate --path ${path}`, { stdio: "inherit" });
        }
        catch (e) {
            console.error("❌ OpenAPI generation failed.");
            return;
        }
        await this.postProcess();
    }
    // Native helper to find json/yaml files recursively
    findSpecs(dir) {
        let results = [];
        const list = fs.readdirSync(dir);
        for (const file of list) {
            const filePath = path.join(dir, file);
            const stat = fs.statSync(filePath);
            if (stat && stat.isDirectory()) {
                results = results.concat(this.findSpecs(filePath));
            }
            else {
                if (file.endsWith(".json") || file.endsWith(".yaml")) {
                    results.push(filePath);
                }
            }
        }
        return results;
    }
}
exports.GenHttpCommand = GenHttpCommand;
